File: daedalus/files/system/opt/daedalus/servers/shell_server.py
#!/usr/bin/env python3
"""Daedalus OS Shell MCP 服务器（白名单与路径验证）。

通过显式的 argv 白名单和路径验证参数提供安全的命令执行。
通过 asyncio.create_subprocess_exec 直接执行子进程（绝不使用 shell=True），
具有净化后的环境变量和 30 秒执行超时。
"""

import asyncio
import json
import logging
import os
import time
from typing import Any, Dict, List, Optional
from mcp.server.fastmcp import FastMCP

mcp = FastMCP("daedalus-shell")

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger("daedalus-shell")

# 默认允许的诊断/只读命令
DEFAULT_ALLOW_COMMANDS = {
    "df",
    "ls",
    "cat",
    "pwd",
    "uname",
    "free",
    "ps",
    "uptime",
    "whoami",
    "ip",
    "arch",
    "hostname",
    "date",
    "ping",
    "systemctl",
}

# 允许通过环境变量覆盖或扩展允许的命令
_env_commands = os.environ.get("ALLOW_COMMANDS")
if _env_commands:
    ALLOW_COMMANDS = set(c.strip() for c in _env_commands.split(",") if c.strip())
else:
    ALLOW_COMMANDS = DEFAULT_ALLOW_COMMANDS

# 路径类参数允许的路径前缀
ALLOWED_PATH_PREFIXES = (
    "/home",
    "/var/log",
    "/tmp",
    "/proc",
    "/sys",
    "/etc/os-release",
    "/usr/lib/os-release",
    "/etc/fedora-release",
    "/etc/almalinux-release",
)

# 显式禁止的敏感路径
BLOCKED_PATHS = (
    "/etc/shadow",
    "/etc/gshadow",
    "/etc/sudoers",
    "/etc/sudoers.d",
    "/root",
)

# 净化后的执行环境（剥离敏感令牌和自定义变量）
CLEAN_ENV = {
    "PATH": "/usr/bin:/bin:/usr/sbin:/sbin",
    "LANG": "C.UTF-8",
}

TIMEOUT_SECONDS = 30.0


def is_path_like(arg: str) -> bool:
    """检查参数字符串是否类似于文件系统路径。"""
    if "\x00" in arg:
        return True
    if arg.startswith("/") or "/" in arg or arg in (".", "..") or arg.startswith(".."):
        return True
    return False


def validate_path(path_str: str) -> str:
    """验证路径参数是否在允许的目录内。

    参数：
        path_str: 要验证的路径字符串。

    返回：
        解析后的规范化路径字符串。

    异常：
        ValueError: 如果检测到空字节。
        PermissionError: 如果路径被阻止或超出允许的目录。
    """
    if "\x00" in path_str:
        raise ValueError("Null bytes are not allowed in path arguments.")

    resolved = os.path.realpath(os.path.abspath(path_str))

    # 检查显式阻止的路径
    for blocked in BLOCKED_PATHS:
        clean_blocked = blocked.rstrip("/")
        if resolved == clean_blocked or resolved.startswith(clean_blocked + "/"):
            raise PermissionError(f"Access to blocked path '{path_str}' ({resolved}) is forbidden.")

    # 检查允许的前缀
    allowed = False
    for prefix in ALLOWED_PATH_PREFIXES:
        clean_prefix = prefix.rstrip("/")
        if resolved == clean_prefix or resolved.startswith(clean_prefix + "/"):
            allowed = True
            break

    if not allowed:
        raise PermissionError(
            f"Path '{path_str}' (resolved: {resolved}) is outside allowed directories: "
            f"{', '.join(ALLOWED_PATH_PREFIXES)}"
        )

    return resolved


def validate_arg(arg: str) -> None:
    """检查参数中的空字节并验证嵌入的路径。"""
    if not isinstance(arg, str):
        raise ValueError(f"Argument must be a string, got {type(arg).__name__}")

    if "\x00" in arg:
        raise ValueError("Null bytes are not allowed in arguments.")

    if "=" in arg and (arg.startswith("-") or arg.startswith("--")):
        _, val = arg.split("=", 1)
        if is_path_like(val):
            validate_path(val)
    elif is_path_like(arg):
        validate_path(arg)


def validate_command(command: str) -> str:
    """根据 ALLOW_COMMANDS 白名单验证命令。"""
    if not command or not isinstance(command, str):
        raise ValueError("Command must be a non-empty string.")

    if "\x00" in command:
        raise ValueError("Null bytes are not allowed in command.")

    cmd_base = os.path.basename(command.strip())
    if cmd_base not in ALLOW_COMMANDS:
        raise PermissionError(f"Command '{command}' is not in ALLOW_COMMANDS allowlist.")

    if "/" in command:
        cmd_dir = os.path.dirname(os.path.realpath(command))
        allowed_bin_dirs = {"/usr/bin", "/bin", "/usr/sbin", "/sbin"}
        if cmd_dir not in allowed_bin_dirs:
            raise PermissionError(f"Command path '{command}' is not in a valid system bin directory.")

    return cmd_base


def _record_audit(
    command: str,
    args: List[str],
    allowed: bool,
    returncode: Optional[int],
    error: Optional[str] = None,
) -> None:
    """记录结构化审计日志条目。"""
    entry = {
        "timestamp": time.time(),
        "tool": "shell_exec",
        "command": command,
        "args": args,
        "allowed": allowed,
        "returncode": returncode,
        "error": error,
    }
    logger.info("AUDIT: %s", json.dumps(entry))
    audit_path = "/var/log/daedalus/audit.jsonl"
    try:
        if os.path.isdir(os.path.dirname(audit_path)):
            with open(audit_path, "a", encoding="utf-8") as f:
                f.write(json.dumps(entry) + "\n")
    except Exception:
        pass


@mcp.tool()
async def shell_exec(command: str, args: List[str] = []) -> Dict[str, Any]:
    """安全地执行白名单中的命令及参数（只读 / 诊断）。

    命令被限制在 argv 白名单中，并且参数路径将针对允许的系统目录
    （/home、/var/log、/tmp、/proc、/sys、/etc/os-release）进行严格验证。
    执行过程在净化后的环境中隔离运行，并具有 30 秒超时限制。

    参数：
        command: 要执行的命令名称（必须在 ALLOW_COMMANDS 中）。
        args: 要传递的命令行参数列表。

    返回：
        包含 stdout、stderr、returncode 以及可选错误信息的字典。
    """
    if args is None:
        args = []

    # 1. 针对白名单验证命令
    try:
        cmd_to_run = validate_command(command)
    except Exception as e:
        _record_audit(command, args, allowed=False, returncode=126, error=str(e))
        return {
            "stdout": "",
            "stderr": f"Command validation failed: {e}",
            "returncode": 126,
            "error": str(e),
        }

    # 2. 验证所有参数
    try:
        for arg in args:
            validate_arg(arg)
    except Exception as e:
        _record_audit(command, args, allowed=False, returncode=126, error=str(e))
        return {
            "stdout": "",
            "stderr": f"Argument validation failed: {e}",
            "returncode": 126,
            "error": str(e),
        }

    # 3. 异步执行子进程（不使用 shell=True）
    try:
        proc = await asyncio.create_subprocess_exec(
            cmd_to_run,
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=CLEAN_ENV,
        )
        try:
            stdout_bytes, stderr_bytes = await asyncio.wait_for(
                proc.communicate(),
                timeout=TIMEOUT_SECONDS,
            )
            stdout_str = stdout_bytes.decode("utf-8", errors="replace")
            stderr_str = stderr_bytes.decode("utf-8", errors="replace")
            _record_audit(command, args, allowed=True, returncode=proc.returncode)
            return {
                "stdout": stdout_str,
                "stderr": stderr_str,
                "returncode": proc.returncode,
            }
        except asyncio.TimeoutError:
            try:
                proc.kill()
                await proc.wait()
            except Exception:
                pass
            err_msg = f"Command timed out after {TIMEOUT_SECONDS} seconds"
            _record_audit(command, args, allowed=True, returncode=124, error=err_msg)
            return {
                "stdout": "",
                "stderr": err_msg,
                "returncode": 124,
                "error": err_msg,
            }
    except Exception as e:
        _record_audit(command, args, allowed=True, returncode=1, error=str(e))
        return {
            "stdout": "",
            "stderr": f"Execution failed: {e}",
            "returncode": 1,
            "error": str(e),
        }


if __name__ == "__main__":
    mcp.run()

File: daedalus/files/system/opt/daedalus/deno/shell_server.ts
/**
 * Daedalus OS Shell MCP 服务器 (Deno)。
 *
 * 为 "daedalus-shell-deno" 实现标准 JSON-RPC 2.0 stdio MCP 服务器。
 * 通过显式参数白名单、路径验证参数、净化执行环境、30 秒超时以及结构化仅追加审计日志，
 * 提供安全的命令执行。
 *
 * 设计为在 Deno 运行时权限边界下运行：
 *   deno run --allow-run=/usr/bin/df,/usr/bin/ls,... shell_server.ts
 */

// 显式允许的诊断/只读命令
export const DEFAULT_ALLOW_COMMANDS = new Set([
  "df",
  "ls",
  "cat",
  "pwd",
  "uname",
  "free",
  "ps",
  "uptime",
  "whoami",
  "ip",
  "arch",
  "hostname",
  "date",
  "ping",
  "systemctl",
]);

// 允许通过环境变量覆盖或扩展允许的命令
const envCommands = Deno.env.get("ALLOW_COMMANDS");
export const ALLOW_COMMANDS: Set<string> = envCommands
  ? new Set(envCommands.split(",").map((c) => c.trim()).filter((c) => c.length > 0))
  : DEFAULT_ALLOW_COMMANDS;

// 路径类参数允许的路径前缀
export const ALLOWED_PATH_PREFIXES = [
  "/home",
  "/var/log",
  "/tmp",
  "/proc",
  "/sys",
  "/etc/os-release",
  "/usr/lib/os-release",
  "/etc/fedora-release",
  "/etc/almalinux-release",
];

// 显式禁止的敏感路径
export const BLOCKED_PATHS = [
  "/etc/shadow",
  "/etc/gshadow",
  "/etc/sudoers",
  "/etc/sudoers.d",
  "/root",
];

// 干净的执行环境
const CLEAN_ENV: Record<string, string> = {
  PATH: "/usr/bin:/bin:/usr/sbin:/sbin",
  LANG: "C.UTF-8",
};

const TIMEOUT_MS = 30000;
const AUDIT_LOG_PATH = Deno.env.get("DAEDALUS_AUDIT_LOG_PATH") || "/var/log/daedalus/audit.jsonl";

/**
 * 判断参数是否类似于文件系统路径。
 */
export function isPathLike(arg: string): boolean {
  if (arg.includes("\0")) {
    return true;
  }
  if (
    arg.startsWith("/") ||
    arg.includes("/") ||
    arg === "." ||
    arg === ".." ||
    arg.startsWith("..")
  ) {
    return true;
  }
  return false;
}

/**
 * 规范化并验证路径参数。
 * 确保路径不触及受阻路径且保持在允许的目录范围内。
 */
export function validatePath(pathStr: string): string {
  if (typeof pathStr !== "string" || pathStr.length === 0) {
    throw new Error("Path must be a non-empty string.");
  }
  if (pathStr.includes("\0")) {
    throw new Error("Null bytes are not allowed in path arguments.");
  }

  // 若路径存在则解析规范 realpath，若不存在则进行规范化
  let resolved: string;
  try {
    resolved = Deno.realPathSync(pathStr);
  } catch {
    // 若路径在磁盘上尚不存在，则相对于 cwd 或根路径解析
    const absolute = pathStr.startsWith("/")
      ? pathStr
      : `${Deno.cwd()}/${pathStr}`;
    resolved = normalizePath(absolute);
  }

  // 检查显式禁止的路径
  for (const blocked of BLOCKED_PATHS) {
    const cleanBlocked = blocked.replace(/\/+$/, "");
    if (resolved === cleanBlocked || resolved.startsWith(cleanBlocked + "/")) {
      throw new Error(`Access to blocked path '${pathStr}' (${resolved}) is forbidden.`);
    }
  }

  // 检查允许的目录前缀
  let allowed = false;
  for (const prefix of ALLOWED_PATH_PREFIXES) {
    const cleanPrefix = prefix.replace(/\/+$/, "");
    if (resolved === cleanPrefix || resolved.startsWith(cleanPrefix + "/")) {
      allowed = true;
      break;
    }
  }

  if (!allowed) {
    throw new Error(
      `Path '${pathStr}' (resolved: ${resolved}) is outside allowed directories: ${ALLOWED_PATH_PREFIXES.join(", ")}`,
    );
  }

  return resolved;
}

/**
 * 针对不存在路径的基础路径规范化工具。
 */
function normalizePath(path: string): string {
  const parts = path.split("/").filter((p) => p.length > 0 && p !== ".");
  const stack: string[] = [];
  for (const part of parts) {
    if (part === "..") {
      stack.pop();
    } else {
      stack.push(part);
    }
  }
  return "/" + stack.join("/");
}

/**
 * 验证单个参数是否包含空字节和嵌入路径。
 */
export function validateArg(arg: string): void {
  if (typeof arg !== "string") {
    throw new Error(`Argument must be a string, got ${typeof arg}`);
  }
  if (arg.includes("\0")) {
    throw new Error("Null bytes are not allowed in arguments.");
  }

  if (arg.includes("=") && (arg.startsWith("-") || arg.startsWith("--"))) {
    const eqIdx = arg.indexOf("=");
    const val = arg.slice(eqIdx + 1);
    if (isPathLike(val)) {
      validatePath(val);
    }
  } else if (isPathLike(arg)) {
    validatePath(arg);
  }
}

/**
 * 依据白名单验证命令名称，并确保无路径遍历行为。
 */
export function validateCommand(command: string): string {
  if (!command || typeof command !== "string") {
    throw new Error("Command must be a non-empty string.");
  }
  if (command.includes("\0")) {
    throw new Error("Null bytes are not allowed in command.");
  }

  const trimmed = command.trim();
  const lastSlash = trimmed.lastIndexOf("/");
  const cmdBase = lastSlash >= 0 ? trimmed.slice(lastSlash + 1) : trimmed;

  if (!ALLOW_COMMANDS.has(cmdBase)) {
    throw new Error(`Command '${command}' is not in ALLOW_COMMANDS allowlist.`);
  }

  if (trimmed.includes("/")) {
    let cmdDir = trimmed.slice(0, lastSlash);
    try {
      cmdDir = Deno.realPathSync(cmdDir);
    } catch {
      // 若 realPathSync 失败则保持 cmdDir 原样
    }
    const allowedBinDirs = new Set(["/usr/bin", "/bin", "/usr/sbin", "/sbin"]);
    if (!allowedBinDirs.has(cmdDir)) {
      throw new Error(`Command path '${command}' is not in a valid system bin directory.`);
    }
  }

  return cmdBase;
}

/**
 * 追加审计日志条目。
 */
export async function recordAudit(
  command: string,
  args: string[],
  allowed: boolean,
  returncode: number | null,
  error?: string,
): Promise<void> {
  const entry = {
    timestamp: Date.now() / 1000,
    tool: "shell_exec",
    command,
    args,
    allowed,
    returncode,
    error: error || null,
  };

  try {
    const encoder = new TextEncoder();
    const line = encoder.encode(JSON.stringify(entry) + "\n");
    // 如果可能则追加写入审计日志
    await Deno.writeFile(AUDIT_LOG_PATH, line, { append: true, create: false });
  } catch {
    // 忽略审计日志写入失败（例如测试期间为只读系统或目录缺失）
  }
}

export interface ShellExecResult {
  stdout: string;
  stderr: string;
  returncode: number;
  error?: string;
}

/**
 * 使用 Deno.Command 安全执行命令。
 */
export async function shellExec(
  command: string,
  args: string[] = [],
): Promise<ShellExecResult> {
  const effectiveArgs = args || [];

  // 1. 验证命令
  let cmdToRun: string;
  try {
    cmdToRun = validateCommand(command);
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    await recordAudit(command, effectiveArgs, false, 126, msg);
    return {
      stdout: "",
      stderr: `Command validation failed: ${msg}`,
      returncode: 126,
      error: msg,
    };
  }

  // 2. 验证所有参数
  try {
    for (const arg of effectiveArgs) {
      validateArg(arg);
    }
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    await recordAudit(command, effectiveArgs, false, 126, msg);
    return {
      stdout: "",
      stderr: `Argument validation failed: ${msg}`,
      returncode: 126,
      error: msg,
    };
  }

  // 3. 使用 Deno.Command 及超时机制执行进程
  try {
    const denoCmd = new Deno.Command(cmdToRun, {
      args: effectiveArgs,
      stdout: "piped",
      stderr: "piped",
      env: CLEAN_ENV,
    });

    const proc = denoCmd.spawn();

    let timeoutId: number | undefined;
    const timeoutPromise = new Promise<never>((_, reject) => {
      timeoutId = (setTimeout(() => {
        try {
          proc.kill("SIGKILL");
        } catch {
          // 进程可能已经结束
        }
        reject(new Error(`Command timed out after ${TIMEOUT_MS / 1000} seconds`));
      }, TIMEOUT_MS) as unknown) as number;
    });

    try {
      const output = await Promise.race([proc.output(), timeoutPromise]);
      if (timeoutId !== undefined) {
        clearTimeout(timeoutId);
      }

      const decoder = new TextDecoder();
      const stdout = decoder.decode(output.stdout);
      const stderr = decoder.decode(output.stderr);
      const returncode = output.code;

      await recordAudit(command, effectiveArgs, true, returncode);
      return {
        stdout,
        stderr,
        returncode,
      };
    } catch (raceErr: unknown) {
      if (timeoutId !== undefined) {
        clearTimeout(timeoutId);
      }
      const msg = raceErr instanceof Error ? raceErr.message : String(raceErr);
      await recordAudit(command, effectiveArgs, true, 124, msg);
      return {
        stdout: "",
        stderr: msg,
        returncode: 124,
        error: msg,
      };
    }
  } catch (execErr: unknown) {
    const msg = execErr instanceof Error ? execErr.message : String(execErr);
    await recordAudit(command, effectiveArgs, true, 1, msg);
    return {
      stdout: "",
      stderr: `Execution failed: ${msg}`,
      returncode: 1,
      error: msg,
    };
  }
}

/**
 * 标准 MCP 工具定义与元数据
 */
const TOOLS = [
  {
    name: "shell_exec",
    description:
      "Execute an allowlisted command with arguments safely (read-only / diagnostic). Restricted to an argv allowlist with strict argument path validation and 30s timeout.",
    inputSchema: {
      type: "object",
      properties: {
        command: {
          type: "string",
          description: "The command name to execute (must be in ALLOW_COMMANDS).",
        },
        args: {
          type: "array",
          items: { type: "string" },
          description: "List of command-line arguments to pass.",
        },
      },
      required: ["command"],
    },
    annotations: {
      readOnlyHint: true,
      destructiveHint: false,
      idempotentHint: true,
      openWorldHint: false,
    },
  },
];

/**
 * JSON-RPC 2.0 stdio MCP 协议处理器
 */
export async function handleMessage(message: Record<string, unknown>): Promise<Record<string, unknown> | null> {
  const jsonrpc = message.jsonrpc;
  const id = message.id;
  const method = message.method;
  const params = (message.params as Record<string, unknown>) || {};

  if (jsonrpc !== "2.0") {
    if (id !== undefined) {
      return {
        jsonrpc: "2.0",
        id,
        error: { code: -32600, message: "Invalid Request: jsonrpc must be '2.0'" },
      };
    }
    return null;
  }

  // 处理通知 / 请求
  switch (method) {
    case "initialize": {
      return {
        jsonrpc: "2.0",
        id,
        result: {
          protocolVersion: "2024-11-05",
          capabilities: {
            tools: {
              listChanged: false,
            },
          },
          serverInfo: {
            name: "daedalus-shell-deno",
            version: "1.0.0",
          },
        },
      };
    }

    case "notifications/initialized":
    case "initialized": {
      return null;
    }

    case "ping": {
      return {
        jsonrpc: "2.0",
        id,
        result: {},
      };
    }

    case "tools/list": {
      return {
        jsonrpc: "2.0",
        id,
        result: {
          tools: TOOLS,
        },
      };
    }

    case "tools/call": {
      const toolName = params.name as string;
      const toolArguments = (params.arguments as Record<string, unknown>) || {};

      if (toolName === "shell_exec") {
        const cmd = String(toolArguments.command || "");
        const rawArgs = toolArguments.args;
        const argsList: string[] = Array.isArray(rawArgs)
          ? rawArgs.map((a) => String(a))
          : [];

        const execResult = await shellExec(cmd, argsList);
        const isError = execResult.returncode !== 0;

        return {
          jsonrpc: "2.0",
          id,
          result: {
            content: [
              {
                type: "text",
                text: JSON.stringify(execResult, null, 2),
              },
            ],
            isError,
          },
        };
      }

      return {
        jsonrpc: "2.0",
        id,
        error: {
          code: -32601,
          message: `Method '${toolName}' not found`,
        },
      };
    }

    default: {
      if (id !== undefined) {
        return {
          jsonrpc: "2.0",
          id,
          error: {
            code: -32601,
            message: `Method '${method}' not found`,
          },
        };
      }
      return null;
    }
  }
}

/**
 * 按行读取 JSON-RPC 消息的主标准输入输出循环。
 */
export async function runServer(): Promise<void> {
  const decoder = new TextDecoder();
  const encoder = new TextEncoder();

  const buf = new Uint8Array(65536);
  let leftover = "";

  while (true) {
    const bytesRead = await Deno.stdin.read(buf);
    if (bytesRead === null) {
      break; // 文件结束 (EOF)
    }

    leftover += decoder.decode(buf.subarray(0, bytesRead));
    const lines = leftover.split("\n");
    leftover = lines.pop() || "";

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }

      try {
        const parsed = JSON.parse(trimmed);
        const response = await handleMessage(parsed);
        if (response) {
          await Deno.stdout.write(encoder.encode(JSON.stringify(response) + "\n"));
        }
      } catch (err: unknown) {
        const errorResponse = {
          jsonrpc: "2.0",
          id: null,
          error: {
            code: -32700,
            message: `Parse error: ${err instanceof Error ? err.message : String(err)}`,
          },
        };
        await Deno.stdout.write(encoder.encode(JSON.stringify(errorResponse) + "\n"));
      }
    }
  }
}

// 直接作为入口点执行时运行服务器
if (import.meta.main) {
  runServer();
}
